Feature/nirmal agnihotri/1017159459 Add Table

NirmalAgnihotri opened the pull request:
### Administrative Info

Monday Board ID: 851212620
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Monday Item ID]-[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes

What changes did you make?
- Created table of agencies
- Implemented searching based on any word in Agency name or number
- Implemented multi-select filtering by storage, staff, status, etc.
- Implemented pagination
- Implemented clickable labels to clear a filter option and clear all button

### Testing
Searched for each item in the table and items not in the table
Changed page numbers and searched/filtered on each page
Filtered for each category
Changed filters with labels, clear all button, and selecting checkboxes
Added new items to database and looked for change on table

- TODO

### Confirmation of Change

Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

Table
![PRTable](https://user-images.githubusercontent.com/53363382/112909197-cbe94c80-90a5-11eb-8731-69b9211b2478.png)

Filtering
![PRFiltering](https://user-images.githubusercontent.com/53363382/112909262-ea4f4800-90a5-11eb-83d3-2eb1461ae5e9.png)

Search
![PRSearch](https://user-images.githubusercontent.com/53363382/112909282-f4714680-90a5-11eb-8ad0-6d750c318067.png)

Pagination
![PRPagination](https://user-images.githubusercontent.com/53363382/112909291-f9ce9100-90a5-11eb-92d2-5c23296d974f.png)

NirmalAgnihotri committed:
Create Table.js

ryanbuui committed:
Fixed some styling issues on CreateAgencyBtn and Navbar components

ryanbuui committed:
Merge branch 'feature/NirmalAgnihotri/1017159459-add-rough-table' of https://github.com/TritonSE/distribution-web-app into feature/NirmalAgnihotri/1017159459-add-rough-table

ryanbuui committed:
Setup basic table component

ryanbuui committed:
Added get route and made an axios request in agency page

NirmalAgnihotri committed:
changed routes to only get table content

NirmalAgnihotri committed:
updated table to contain all data from database

NirmalAgnihotri committed:
added some default table styling

ryanbuui committed:
Converted axios request into fetch api for Agency

ryanbuui committed:
Added filter component

NirmalAgnihotri committed:
added pagination

ryanbuui committed:
Updated Agency Table and styled

ryanbuui committed:
Small updates to the styling of agency table

NirmalAgnihotri committed:
added last and first pagination

NirmalAgnihotri committed:
added multiselect

ryanbuui committed:
Updated filtering functionality for agency table

NirmalAgnihotri committed:
multiselect filters multiple

NirmalAgnihotri committed:
changed filter objects from arrays to higher level objects

NirmalAgnihotri committed:
added multiselect filtering and dynamic filter options

NirmalAgnihotri committed:
changed data in tableContent and styled table

NirmalAgnihotri committed:
Merge branch 'main' into feature/NirmalAgnihotri/1017159459-add-rough-table

NirmalAgnihotri committed:
created env

NirmalAgnihotri committed:
optimized search, added selected filter buttons

NirmalAgnihotri committed:
fixed select dropdown appearing over filter items

NirmalAgnihotri committed:
added joined date, transportation, and storage to table schema

wllmwu committed:
Update schema preparation in AgencyProfileForm

wllmwu committed:
Remove .env

NirmalAgnihotri committed:
Merge branch 'main' into feature/NirmalAgnihotri/1017159459-add-rough-table

NirmalAgnihotri committed:
changed schema, added joined in filtering

NirmalAgnihotri committed:
added transportation and joined in filtering

NirmalAgnihotri committed:
style changes and transport filtering

ecflores30 committed:
Merge branch 'feature/NirmalAgnihotri/1017159459-add-rough-table' into main

ryanbuui committed:
Updated styling for filter options and checkboxes on agency table

NirmalAgnihotri committed:
changed status, region, staff to upeercase

NirmalAgnihotri committed:
changed css for x's

NirmalAgnihotri committed:
changed width of transportation

NirmalAgnihotri committed:
increased width for transportation box

NirmalAgnihotri committed:
added comments, pagination to clear all

NirmalAgnihotri committed:
added comments for css

NirmalAgnihotri committed:
changed == to ===

NirmalAgnihotri committed:
update dropdown to close on click away

NirmalAgnihotri committed:
code structure and style changes

NirmalAgnihotri committed:
more styling and spacing fixes

NirmalAgnihotri committed:
Merge branch 'main' into feature/NirmalAgnihotri/1017159459-add-rough-table

NirmalAgnihotri committed:
changed affected font sizes after merge

NirmalAgnihotri committed:
changed to 20 per page

NirmalAgnihotri committed:
gray out pagination when disabled

sblumin reviewed (changes requested):
Please check out my comments and address them!

wllmwu commented:
Note also--when you refresh the table page, it goes to displaying raw JSON:

![Screen Shot 2021-03-30 at 11 39 42](https://user-images.githubusercontent.com/74469686/113039600-b1b67980-914c-11eb-862f-881a9ecadab5.png)

This is because when you refresh, the browser manually sends a GET request to the server using whatever route is in the address bar, bypassing React Router which is purely on the frontend. A while ago, we added this catch-all route in backend/app.js to solve this:
```
// Catch-all route
app.get('/*', (req, res, next) => {
  res.sendFile(path.join(__dirname, 'build/index.html'), (err) => {
    if (err) {
      next();
    }
  });
});
```
This just sends back index.html, containing all the React Router logic, for any request that doesn't match another route. Then React Router will navigate itself back to the correct page and all is good.

However, because the route to your table page on the frontend, `/agency`, also matches the backend GET route `/agency/` that you are using to get all agencies, the backend executes that route when the browser refreshes, and sends back the raw JSON instead of falling through to the catch-all. It is this route at the bottom of backend/routes/agency.js:
```
router.get("/", async (req, res, next) => {
  try {
    const agency = await Agency.find({}).select("tableContent");
    ...
```

I think the solution is to change the name of your route on the backend so it doesn't match. In the above snippet, change it to `router.get("/all", async ...)` or something, and change the fetch on the frontend too.

P.S. This is another reason why we will have to use the agency ID in the route to the agency profile, instead of passing it through props. If you refresh the page for whatever reason, the props are lost.

sblumin commented:
> Note also--when you refresh the table page, it goes to displaying raw JSON:
> 
> ![Screen Shot 2021-03-30 at 11 39 42](https://user-images.githubusercontent.com/74469686/113039600-b1b67980-914c-11eb-862f-881a9ecadab5.png)
> 
> This is because when you refresh, the browser manually sends a GET request to the server using whatever route is in the address bar, bypassing React Router which is purely on the frontend. A while ago, we added this catch-all route in backend/app.js to solve this:
> 
> ```
> // Catch-all route
> app.get('/*', (req, res, next) => {
>   res.sendFile(path.join(__dirname, 'build/index.html'), (err) => {
>     if (err) {
>       next();
>     }
>   });
> });
> ```
> 
> This just sends back index.html, containing all the React Router logic, for any request that doesn't match another route. Then React Router will navigate itself back to the correct page and all is good.
> 
> However, because the route to your table page on the frontend, `/agency`, also matches the backend GET route `/agency/` that you are using to get all agencies, the backend executes that route when the browser refreshes, and sends back the raw JSON instead of falling through to the catch-all. It is this route at the bottom of backend/routes/agency.js:
> 
> ```
> router.get("/", async (req, res, next) => {
>   try {
>     const agency = await Agency.find({}).select("tableContent");
>     ...
> ```
> 
> I think the solution is to change the name of your route on the backend so it doesn't match. In the above snippet, change it to `router.get("/all", async ...)` or something, and change the fetch on the frontend too.
> 
> P.S. This is another reason why we will have to use the agency ID in the route to the agency profile, instead of passing it through props. If you refresh the page for whatever reason, the props are lost.

Thanks for bringing this up, please address this also!!!

emmanuelclflores commented on frontend/src/components/AgencyTable/DataTable.css:
```diff
@@ -0,0 +1,51 @@
+.table {
```
I second this

emmanuelclflores commented on frontend/src/pages/Agency/Agency.css:
```diff
@@ -0,0 +1,82 @@
+.search-container {
```
Could you implement some additional comments here to group some of the styles?

ecflores30 reviewed (changes requested):
Looks great! Just be sure to implement some of the headers/minor stylistic changes that Sara mentioned.

NirmalAgnihotri committed:
code style and structure changes

NirmalAgnihotri committed:
route change

NirmalAgnihotri committed:
route changes

NirmalAgnihotri committed:
added comments and method headers

ryanbuui committed:
Added comments to a majority of css files

NirmalAgnihotri committed:
added comments and fixed route

NirmalAgnihotri committed:
Merge branch 'feature/NirmalAgnihotri/1017159459-add-rough-table' of https://github.com/TritonSE/Distribution-Web-App into feature/NirmalAgnihotri/1017159459-add-rough-table

NirmalAgnihotri committed:
Added file headers

NirmalAgnihotri committed:
Removed eslint disable

NirmalAgnihotri committed:
Create package-lock.json

NirmalAgnihotri committed:
fixed failing navbar circle CI test

NirmalAgnihotri commented:
> > Note also--when you refresh the table page, it goes to displaying raw JSON:
> > ![Screen Shot 2021-03-30 at 11 39 42](https://user-images.githubusercontent.com/74469686/113039600-b1b67980-914c-11eb-862f-881a9ecadab5.png)
> > This is because when you refresh, the browser manually sends a GET request to the server using whatever route is in the address bar, bypassing React Router which is purely on the frontend. A while ago, we added this catch-all route in backend/app.js to solve this:
> > ```
> > // Catch-all route
> > app.get('/*', (req, res, next) => {
> >   res.sendFile(path.join(__dirname, 'build/index.html'), (err) => {
> >     if (err) {
> >       next();
> >     }
> >   });
> > });
> > ```
> > 
> > 
> > This just sends back index.html, containing all the React Router logic, for any request that doesn't match another route. Then React Router will navigate itself back to the correct page and all is good.
> > However, because the route to your table page on the frontend, `/agency`, also matches the backend GET route `/agency/` that you are using to get all agencies, the backend executes that route when the browser refreshes, and sends back the raw JSON instead of falling through to the catch-all. It is this route at the bottom of backend/routes/agency.js:
> > ```
> > router.get("/", async (req, res, next) => {
> >   try {
> >     const agency = await Agency.find({}).select("tableContent");
> >     ...
> > ```
> > 
> > 
> > I think the solution is to change the name of your route on the backend so it doesn't match. In the above snippet, change it to `router.get("/all", async ...)` or something, and change the fetch on the frontend too.
> > P.S. This is another reason why we will have to use the agency ID in the route to the agency profile, instead of passing it through props. If you refresh the page for whatever reason, the props are lost.
> 
> Thanks for bringing this up, please address this also!!!

Changed the backend route name to /table/all

NirmalAgnihotri committed:
fixed typos

sblumin commented on frontend/src/components/AgencyTable/Dropdown.js:
```diff
@@ -0,0 +1,61 @@
+import React, { useState } from 'react';
```
these single quotes need to be updated

sblumin commented on frontend/src/pages/Agency/Agency.js:
```diff
@@ -1,12 +1,25 @@
+import { getJWT } from "../../auth";
+
+/**
+ * The AgencyTable component is the main component for the talbe and contains the filtering
```
talbe --> table,
also in the header for foptions, should it be fOptions?

sblumin commented on frontend/src/pages/Agency/Agency.js:
```diff
@@ -82,122 +100,135 @@ function AgencyTable() {
+   * @param {Object} row JSON object that contains the data for the current row being checked against the set filter options
+   * @returns Boolean value, true if the current row's data matches the filter options and false otherwise
+   */
+  function checkOptions(row){
```
For styling consistency, please add spaces in between text and the opening brackets, for example `function checkOptions(row){` should be `function checkOptions(row) {` and so on. Please see if you can do so for this file and anywhere else that may need it!

sblumin reviewed (changes requested):
Looks great, I noticed a couple typos/style stuff but minor changes!

emmanuelclflores commented on frontend/src/components/AgencyTable/DataTable.jsx:
```diff
@@ -0,0 +1,42 @@
+        </tr>
+      </thead>
+       <tbody>
+         { data.map(dat =>
```
Did you mean "data" as opposed to "dat?"

NirmalAgnihotri replied:
dat is just the temp variable I picked for each element in data since the map function requires it.

NirmalAgnihotri committed:
minor code style and typo fixes

NirmalAgnihotri committed:
bracket spacing

sblumin commented on frontend/src/pages/Agency/Agency.js:
```diff
@@ -63,17 +76,22 @@ function AgencyTable() {
+        if(!(filters["Joined In"].hasOwnProperty(dat.tableContent.dateOfInitialPartnership.substring(6)))) {
           let year = dat.tableContent.dateOfInitialPartnership.substring(6);
           filters["Joined In"][year] = false;
         }
```
Not sure if it's just hard to tell but is there a missing closing brace here? Also, why are the two if statements on 93 and 94 stacked on top of each other? It seems like they should be nested and it looks like one of the if statements is missing a closing brace. Can you confirm for me if this is correct?

NirmalAgnihotri replied:
Just indented line 94 to fix the indentation. The closing brace was there but because the indentation was a bit messed up, it was hard to tell.

NirmalAgnihotri committed:
indentation fix

NirmalAgnihotri committed:
misaligned braces fix

NirmalAgnihotri committed:
indentation in select.js

sblumin reviewed (approved):
LGTM! Great job

ecflores30 reviewed (approved):
LGTM! Well done!
